### DISCUSSION TITLE
Errors on Running the simulation

### DISCUSSION THREAD

**Discussion by JauJieYou** [surgical-robotics-ai/surgical_robotics_challenge#69] [Q&A]
I am newbie to AMBF, and I suffered from erros while [Running the simulation](https://github.com/collaborative-robotics/surgical_robotics_challenge). When I run the code `~/ambf/bin/lin-x86_64/ambf_simulator --launch_file ~/ambf/surgical_robotics_challenge/launch.yaml -l 0,1,3,4,14,15 -p 120 -t 1 --override_max_comm_freq 120` 
There are two problem I cannot solve it after google
![Screenshot from 2023-04-07 14-59-29](https://user-images.githubusercontent.com/129126407/230564025-a953f68a-a6ed-4f1b-80eb-79ea5e4eb30f.png)
1) there was no light? The simulator window are dark, though something inside could be seen.
`
WARNING! For File "./ADF/world/world_stereo.yaml", ADF version not defined thus assuming VERSION_1_0
WARNING! For File "./ADF/input_devices/input_devices.yaml", ADF version not defined thus assuming VERSION_1_0
INFO! INITIALIZING ROS NODE HANDLE
...
INFO! CAMERA "cameraL" MONITOR NUMBER "1" IS NOT IN RANGE OF AVAILABLE MONITORS "1", USING DEFAULT
INFO! Thread Joined: cameraL
INFO! CAMERA "cameraR" MONITOR NUMBER "2" IS NOT IN RANGE OF AVAILABLE MONITORS "1", USING DEFAULT
INFO! Thread Joined: cameraR
`
2) The bilateral PSM looks are there, though it is dark scene. But there was no movement as examples. It showed
`...
ERROR! COLLISION MESH TYPE "" NOT UNDERSTOOD 
INFO! Thread Joined: baselink
INFO! Thread Joined: toolyawlink
WARNING! COMMUNICATION TYPE FOR OBJECT NAMED baselink-yawlink OF TYPE: JOINT NOT IMPLEMENTED YET. IGNORING.
WARNING! COMMUNICATION TYPE FOR OBJECT NAMED yawlink-pitchbacklink OF TYPE: JOINT NOT IMPLEMENTED YET. IGNORING.
...
WARNING! Required parent name of /ambf/env/lights/light2 set as BODY CameraFrame. Parenting to closest match /ambf/env/BODY CameraFrame
WARNING! Required parent name of /ambf/env/cameras/cameraL set as BODY CameraFrame. Parenting to closest match /ambf/env/BODY CameraFrame
...
`
What is my mistake, and how can I solve this problem?

**Comment by adnanmunawar** [comment]
Hello,

Thanks for questions. Here are the answers to them.

1. The lighting is dark indeed, and that's what we set it do. When you are closer to the anatomy, you should be able to see and when are further, as in the picture above, thinks appear darker. Don't worry about the warnings, these are also normal. This is GIF of how the scene should look naturally and it appears that your's look the same.
![Simulator Lighting](https://user-images.githubusercontent.com/5005445/230647087-21802d86-fed2-465b-832b-8eee7be8923b.gif)


2. To control the arms, you have to launch the control scripts in the scripts folder after launching the scene. There are many examples but it is easiest to use this one (https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/scripts/surgical_robotics_challenge/examples/gui_based_control.py) and this would bring up a GUI to control the pose of different PSMs and the Camera. Again, don't worry about the warnings in the console, these specific ones are expected.

**Reply by JauJieYou** [reply]
Dear Adnan
Thank you very much. 
![Screenshot from 2023-04-10 15-57-09](https://user-images.githubusercontent.com/129126407/230856643-bf3ea744-a54c-49d0-80e6-3777c736c290.png)
But I still cannot find the automation script, as on the web gif. Is there some example?

**Reply by adnanmunawar** [reply]
Hi, I added a link to that script, it is located in the scripts/surgical_robotics_challenge/examples folder. Is that what you are asking?

**Reply by JauJieYou** [reply]
Dear Adnan
  Is there a script as your web example 
![figure_eight](https://user-images.githubusercontent.com/129126407/232710677-10994bbb-1633-4f72-9f7a-667275604108.gif)

I don't know how to write such a script. Is there possible of offer the teaching script? 
Thanks a lot

**Reply by adnanmunawar** [reply]
This video was created by tele-operating the PSMs via MTMs and not a script. If you have access to the dVRK MTMs you can do the same. If not, we can probably re-record this and share a ROS Bag with you with the trajectories for the PSMs.

**Comment by sunshineinsandiego** [comment]
Thanks Adnan! Any chance you could share that ROS Bag? Would definitely be interested.

**Reply by adnanmunawar** [reply]
Of course, @ruiruihuangannie do you think we could perform the task 2 and 3 with the first environment and record two ROS bags?

**Reply by sunshineinsandiego** [reply]
Terrific, thank you!  And once those ROS bags are available, do I then run `launch_crtk_interface.py` to visualize the actions in the simulator?

**Reply by ruiruihuangannie** [reply]
@adnanmunawar Will do!

**Reply by JauJieYou** [reply]
Thank you very much, It will be very interesting.

**Reply by ruiruihuangannie** [reply]
@adnanmunawar I uploaded the two ROS bags to google folder: https://drive.google.com/drive/folders/1mJzqyKD0H4i4XwSC1BkflCaNh1wkgrEf?usp=sharing. Please let me know if i can be of any other help :)

**Comment by sunshineinsandiego** [comment]
Thanks all for your help! Any chance you have advice on actually running these ROS bags in the simulator? Is it simply a matter of starting `launch_crtk_interface.py` and then playing the ROS bags?

**Reply by adnanmunawar** [reply]
Hello, @ruiruihuangannie can you please tell us what ROS topics did you bag? In the past we recorded the lowest level ROS commands for AMBF, i.e. the `ambf_msgs/RigidBodyCommand` and `ambf_msgs/ActuatorCommand`. With those collected, we simply had to run the simulator and play the ROS bag, without needing any other scripts.

**Reply by ruiruihuangannie** [reply]
Hi!

When I recorded the rosbags, I directly subscribed to four AMBF topics without launching the crtk interface. The four topics are: 
- /ambf/env/psm1/baselink/Command
- /ambf/env/psm2/baselink/Command
- /ambf/env/psm1/Actuator0/Command
- /ambf/env/psm2/Actuator0/Command

I recorded them using the command `rosbag record -O ambf_data.bag [the name of the four topics]`

If I understand correctly (and please correct me if I am wrong), I think you are asking about how to play back the ROS bags that are already collected. For me, I always use the command` rosbag play [the name of the ROS bag]` to playback the bag files. One thing to be aware of is that the scene and the rigid bodies must be reset before playing back the ROS bags.

I hope this helps, and please let me know if you have more questions! I will remember to check the discussions :)
Annie

**Reply by sunshineinsandiego** [reply]
Thanks Annie. My question is not about playing back the ROS bag (yes, `rosbag play` is great). But I don't know how to make sure that the topics which are being played / transmitted are being received by the PSM arms in the simulator.

**Reply by ruiruihuangannie** [reply]
Hi!

When you replay the topics while opening the simulator in another terminal, you should be able to see the trajectory being replayed by the PSM arms. 

The reasoning is that when you replay the ROS bags, a sequence of commanded states (identical to those generated using the physical MTM console) is published to the aforementioned topics. Therefore, you should see the PSM moving inside the simulator without writing additional scripts.

Regards,
Annie

**Comment by sunshineinsandiego** [comment]
Awesome, thank you Annie!

**Comment by synvialfajrine** [comment]
Hi, I also have a problem when run the same simulation. I use this `./ambf_simulator --launch_file <surgical_robotics_challenge>/launch.yaml -l 0,1,3,4,14,15 -p 120 -t 1 --override_max_comm_freq 120` and the simulation only shows white windows.

**Reply by adnanmunawar** [reply]
Hello, please read the instructions again. The `<surgical_robotics_challenge>` part needs to be replaced with the location of where you cloned this repo.
